examples/tcp-window-clamps.bpf.c

```diff
@@ -0,0 +1,90 @@
+#include <vmlinux.h>
+#include <bpf/bpf_tracing.h>
+#include <bpf/bpf_core_read.h>
+#include "maps.bpf.h"
+
+/* Minimum value for tp->rcv_ssthresh that is not considered a clamp */
+#define MIN_CLAMP 32 * 1024
+
+struct {
+    __uint(type, BPF_MAP_TYPE_ARRAY);
+    __uint(max_entries, 1);
+    __type(key, u32);
+    __type(value, u64);
+} tcp_window_clamps_total SEC(".maps");
+
+static int handle_tcp_sock(struct tcp_sock *tp)
+{
+    u32 zero = 0, rcv_ssthresh;
+
+    if (!tp) {
+        return 0;
+    }
+
+    rcv_ssthresh = BPF_CORE_READ(tp, rcv_ssthresh);
+
+    if (rcv_ssthresh < MIN_CLAMP) {
+        increment_map(&tcp_window_clamps_total, &zero, 1);
+    }
+
+    return 0;
+}
+
+#ifdef FENTRY_SUPPORT
+// If fentry/fexit is supported, use it for simpler and faster probe.
+// You need to pass -DFENTRY_SUPPORT in compiler flags to enable this.
+
+SEC("fexit/tcp_try_rmem_schedule")
+int BPF_PROG(tcp_try_rmem_schedule_exit, struct sock *sk)
+{
+    return handle_tcp_sock((struct tcp_sock *) sk);
+}
+
+#else
+// Otherwise, fall back to good old kprobe.
+
+struct {
+    __uint(type, BPF_MAP_TYPE_LRU_HASH);
+    __uint(max_entries, 1024);
+    __type(key, u64);
+    __type(value, struct sock *);
+} tcp_rmem_schedule_enters SEC(".maps");
+
+static int enter_key()
+{
+    u64 pid = bpf_get_current_pid_tgid();
+    if (pid) {
+        return pid;
+    }
+
+    return bpf_get_smp_processor_id();
+}
+
+SEC("kprobe/tcp_try_rmem_schedule")
+int BPF_KPROBE(tcp_try_rmem_schedule, struct sock *sk)
+{
+    u64 key = enter_key();
+
+    bpf_map_update_elem(&tcp_rmem_schedule_enters, &key, &sk, BPF_NOEXIST);
+
+    return 0;
+}
+
+SEC("kretprobe/tcp_try_rmem_schedule")
```

Comment:
I met performance issue when use kreprobe to probe function ipt_do_table() in kernel 5.4.0.   So I'd suggest that you'd better have a LNP test before this enabled on production.

Reply:
What do you mean by LNP test?

Reply:
load and performance, stress test

Reply:
That's a given. I did a quick test with kernel source download in a VM:

```
ivan@vm:~$ curl -o /dev/null https://cdn.kernel.org/pub/linux/kernel/v6.x/linux-6.0.11.tar.xz
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100  127M  100  127M    0     0  44.5M      0  0:00:02  0:00:02 --:--:-- 44.5M
```

I see ~4k packets going via iptables for this:

```
[3973:134454161] -A INPUT -i eth0 -p tcp -m tcp --sport 443 -j ACCEPT
```

There are 1286 bpf program runs and ~8ms of CPU time spent:

```
# HELP ebpf_exporter_ebpf_program_info Info about ebpf programs
# TYPE ebpf_exporter_ebpf_program_info gauge
ebpf_exporter_ebpf_program_info{config="tcp-window-clamps",id="552",program="tcp_try_rmem_schedule",tag="f14b021593f58e05"} 1
ebpf_exporter_ebpf_program_info{config="tcp-window-clamps",id="553",program="tcp_try_rmem_schedule_ret",tag="d88afa963de02adb"} 1

# HELP ebpf_exporter_ebpf_program_run_count_total How many times has the program been executed
# TYPE ebpf_exporter_ebpf_program_run_count_total counter
ebpf_exporter_ebpf_program_run_count_total{id="552"} 1286
ebpf_exporter_ebpf_program_run_count_total{id="553"} 1286

# HELP ebpf_exporter_ebpf_program_run_time_seconds How long has the program been executing
# TYPE ebpf_exporter_ebpf_program_run_time_seconds counter
ebpf_exporter_ebpf_program_run_time_seconds{id="552"} 0.00434047
ebpf_exporter_ebpf_program_run_time_seconds{id="553"} 0.003490548
```

With `fexit` based probe it drops to 1190 runs and 3ms of CPU time:

```
# HELP ebpf_exporter_ebpf_program_run_count_total How many times has the program been executed
# TYPE ebpf_exporter_ebpf_program_run_count_total counter
ebpf_exporter_ebpf_program_run_count_total{id="547"} 1190

# HELP ebpf_exporter_ebpf_program_run_time_seconds How long has the program been executing
# TYPE ebpf_exporter_ebpf_program_run_time_seconds counter
ebpf_exporter_ebpf_program_run_time_seconds{id="547"} 0.002882893
```

The number of runs depends on how buffers are drained.

With curl spending 711ms of combined system and user time it comes down to 1.13% for `kprobe` + `kretprobe` and 0.42% for `fexit`. Whether that's an acceptable overhead is up to any consumer (as with any other config).

Reply:
@bobrik Thanks for the data. 
In old kernel versions(at least 5.4.0),  kretprobe is implemented with acquire global locks raw_spin_lock_irqsave(&rp->lock, flags) before get a free instance in pre_handler_kretprobe(), so when there are high traffics (different 5 tuples) handled by different CPUs, it may hit this issue. You can see high si  usage and high latency of tcp packets.